### DISCUSSION TITLE
qtile-contrib idea

### DISCUSSION THREAD

**Discussion by mbenford** [qtile/qtile#2205] [Ideas]
Have you guys ever entertained the idea of having a secondary repo called `qtile-contrib`, where smaller contributions from the community that might not "belong" in the main repo could go? I'm talking about small tweaks, specialized widgets, simple automation, etc. that probably aren't core to qtile and would end up bloating the main codebase.

I've been using qtile for over a month and have coded a few things to improve my personal workflow, but I don't feel like bundling them in a PR because they are either too simple for someone else to replicate or too specialized. But, they are done and working, nonetheless, and perhaps they could be useful to other people without them having to stumble upon my dotfiles. :)

**Comment by tych0** [comment]
We have https://github.com/qtile/qtile-examples which people put these kinds of things in.

**Comment by mbenford** [comment]
`qtile-examples` is great, but I was thinking about something like a package that people could `pip install` and use, much like [xmonad-contrib](https://hackage.haskell.org/package/xmonad-contrib).

**Comment by m-col** [comment]
FWIW I think this is a good idea, and if it were to go ahead I'd like to see some things from libqtile moved there, namely `Popup` and ideally the extensions. I also have a couple other bits and pieces that I could add (e.g. the graphical notification manager would be a good fit there).

It might also be a nice place to put any new widgets, and to pass maintenance responsibility to those that submit them.

**Comment by tych0** [comment]
I like the idea of making contributions easier. I'm a little worried about how it will work in practice, though. People regularly complain that qtile-examples is out of date, and we offer it as an explicitly unmaintained repository of ideas.

If we offer a maintained (i.e. installable via pip) thing, we would definitely need to test it, since we like to break our API on a regular basis. Coordinating merges of API breaks across repos would be annoying, which suggests it should live in the main repo as a 'contrib' module. But if it's just a submodule in qtile, why not just merge useful tidbits into core?

If we decide we don't want to have these sorts of things live in the core, an alternative would be to have a `libqtile.contrib` submodule where people can put this stuff. I'd want two things: 1. people should write tests for their contrib stuff to ensure it doesn't break and 2. we reserve the right to delete stuff from contrib at any time.

An alternative would be wait until someone has time to do an implementation of `qtile migrate` from #1773. I'm hopeful that will be the path forward to stabilizing our API.

**Comment by m-col** [comment]
Yeah, fair enough. It does seem like we'd be in for a world of pain no matter which way we did something. It seems the primary issues are these tidbits breaking due to deprecations, and the complaints we would receive when this inevitably does happen. Like you said, testing and `qtile migrate` will help reduce the deprecation. We'd also ideally have blaringly obvious disclaimers that none of that stuff is maintained by us.

This makes me wonder... maybe something like a submodule or git repository is too concrete, whereas something like github discussions might be a more quick-and-easy solution, where breakages would be detected by users of the snippets (and broken things that nobody use can be forgotten). It'd also be trivial for any randomer with a github account to post a reply with a fixed version, meaning nobody has to give it any attention bar those that use it.
